From 5364d1ccba61ac0a75247251b509b26860ca4358 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Wed, 24 Dec 2014 19:38:40 +0700
Subject: [PATCH 1/5] Revert "XDivert changes for MSIM"

This reverts commit fa66d7833488cff0c0c6a69c92e87e2fc78af894.

XDivert links to some package we don't have, leading to a crash.

Change-Id: I740594dfbb1700fa45bf92dcdad4d23a03229b72
---
 res/values/strings.xml                         |  6 ------
 res/xml/call_feature_setting_msim.xml          | 10 ----------
 src/com/android/phone/CallFeaturesSetting.java | 10 ----------
 3 files changed, 26 deletions(-)

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 40a0aca..8fbc2eb 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -1477,12 +1477,6 @@
     <string name="exportAllcontatsNoContacts"> No Contacts found to export to SIM !!!</string>
     <string name="cursorError"> Cursor is not having proper data !!! </string>
 
-     <!-- XDivert Settings -->
-     <!-- Title for the XDivert Feature displayed in Call Settings -->
-     <string name="xdivert_title">X-Divert</string>
-     <!-- Summary for the XDivert Feature displayed in Call Settings -->
-     <string name="xdivert_summary">CFNRc Settings</string>
-
     <!-- UPLMN Prefer -->
     <string name="uplmn_list_setting_title">Preferred networks</string>
     <string name="network_id">Network ID</string>
diff --git a/res/xml/call_feature_setting_msim.xml b/res/xml/call_feature_setting_msim.xml
index 0db5700..213f0a6 100644
--- a/res/xml/call_feature_setting_msim.xml
+++ b/res/xml/call_feature_setting_msim.xml
@@ -79,16 +79,6 @@
         android:persistent="true"
         android:defaultValue="true" />
 
-      <PreferenceScreen
-          android:key="button_xdivert"
-          android:title="@string/xdivert_title"
-          android:summary="@string/xdivert_summary"
-          android:persistent="false">
-          <intent android:action="android.intent.action.MAIN"
-                  android:targetPackage="com.qti.xdivert"
-                      android:targetClass="com.qti.xdivert.XDivertSetting" />
-      </PreferenceScreen>
-
     <!-- Add videocall settings menu  -->
     <PreferenceCategory
         android:key="button_videocall_settings_key"
diff --git a/src/com/android/phone/CallFeaturesSetting.java b/src/com/android/phone/CallFeaturesSetting.java
index 0ac7ae6..4cadc7a 100644
--- a/src/com/android/phone/CallFeaturesSetting.java
+++ b/src/com/android/phone/CallFeaturesSetting.java
@@ -191,7 +191,6 @@ public class CallFeaturesSetting extends PreferenceActivity
             "phone_account_settings_preference_screen";
 
     private static final String BUTTON_SELECT_SUB_KEY  = "button_call_independent_serv";
-    private static final String BUTTON_XDIVERT_KEY = "button_xdivert";
 
     private Intent mContactListIntent;
 
@@ -1754,15 +1753,6 @@ public class CallFeaturesSetting extends PreferenceActivity
                     "com.android.phone.msim.MSimCallFeaturesSubSetting");
         }
 
-        if (isMsim && TelephonyManager.getDefault().getMultiSimConfiguration() !=
-                TelephonyManager.MultiSimVariants.DSDS) {
-            PreferenceScreen mXDivertPref = (PreferenceScreen) findPreference(BUTTON_XDIVERT_KEY);
-            if (mXDivertPref != null) {
-                log("Remove xdivert preference");
-                prefSet.removePreference(mXDivertPref);
-            }
-        }
-
         if (mButtonDTMF != null) {
             int dtmf = Settings.System.getInt(getContentResolver(),
                     Settings.System.DTMF_TONE_TYPE_WHEN_DIALING, Constants.DTMF_TONE_TYPE_NORMAL);
-- 
1.9.3 (Apple Git-50)


From b04ddd209c068a778ece0b81b45c8f95df2d2c37 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Tue, 23 Dec 2014 22:15:37 +0700
Subject: [PATCH 2/5] Telephony: fix 'up' navigation for SelectSubscription

The 'up' command should be passed down to the activity in the tab

Change-Id: I70d5e4d5cb2d44f62cca481337f108d122726e14
---
 src/com/android/phone/msim/MSimCallFeaturesSubSetting.java | 6 +-----
 src/com/android/phone/msim/SelectSubscription.java         | 9 +++++++++
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/src/com/android/phone/msim/MSimCallFeaturesSubSetting.java b/src/com/android/phone/msim/MSimCallFeaturesSubSetting.java
index c0770b7..c23f666 100644
--- a/src/com/android/phone/msim/MSimCallFeaturesSubSetting.java
+++ b/src/com/android/phone/msim/MSimCallFeaturesSubSetting.java
@@ -1991,11 +1991,7 @@ public class MSimCallFeaturesSubSetting extends PreferenceActivity
     public boolean onOptionsItemSelected(MenuItem item) {
         final int itemId = item.getItemId();
         if (itemId == android.R.id.home) {  // See ActionBar#setDisplayHomeAsUpEnabled()
-            Intent intent = new Intent();
-            intent.setClassName(UP_ACTIVITY_PACKAGE, UP_ACTIVITY_CLASS);
-            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-            startActivity(intent);
-            finish();
+            CallFeaturesSetting.goUpToTopLevelSetting(this);
             return true;
         }
         return super.onOptionsItemSelected(item);
diff --git a/src/com/android/phone/msim/SelectSubscription.java b/src/com/android/phone/msim/SelectSubscription.java
index 7a2b4e7..ab858d4 100644
--- a/src/com/android/phone/msim/SelectSubscription.java
+++ b/src/com/android/phone/msim/SelectSubscription.java
@@ -29,6 +29,7 @@
 
 package com.android.phone.msim;
 
+import android.app.Activity;
 import android.app.TabActivity;
 import android.content.Intent;
 import android.os.Bundle;
@@ -36,6 +37,7 @@ import android.telephony.SubInfoRecord;
 import android.telephony.SubscriptionManager;
 import android.telephony.TelephonyManager;
 import android.util.Log;
+import android.view.MenuItem;
 import android.widget.TabHost;
 import android.widget.TabHost.TabSpec;
 
@@ -102,6 +104,13 @@ public class SelectSubscription extends  TabActivity {
     }
 
     @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        String tabTag = getTabHost().getCurrentTabTag();
+        Activity activity = getLocalActivityManager().getActivity(tabTag);
+        return activity.onOptionsItemSelected(item);
+    }
+
+    @Override
     protected void onResume() {
         super.onResume();
     }
-- 
1.9.3 (Apple Git-50)


From 7ca03b5f7109a0feced711a206ea4dcfb88b8798 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Wed, 10 Dec 2014 14:30:28 +0700
Subject: [PATCH 3/5] Fix Quick Settings for MSim [2/2]

- Use active data sim for setDataEnabled/getDataEnabled
- Also update system setting  for setDataEnabled so that the correct
  state is displayed in Settings

Change-Id: I1c72d9955b1d068ef276386b761f7aa2d664fdc2
---
 src/com/android/phone/PhoneInterfaceManager.java | 23 +++++++++++++++--------
 1 file changed, 15 insertions(+), 8 deletions(-)

diff --git a/src/com/android/phone/PhoneInterfaceManager.java b/src/com/android/phone/PhoneInterfaceManager.java
index 6c23a8e..3ec936b 100644
--- a/src/com/android/phone/PhoneInterfaceManager.java
+++ b/src/com/android/phone/PhoneInterfaceManager.java
@@ -1146,17 +1146,13 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
 
     // FIXME: subId version needed
     public boolean enableDataConnectivity() {
-        enforceModifyPermission();
-        long subId = SubscriptionManager.getDefaultDataSubId();
-        getPhone(subId).setDataEnabled(true);
+        setDataEnabled(true);
         return true;
     }
 
     // FIXME: subId version needed
     public boolean disableDataConnectivity() {
-        enforceModifyPermission();
-        long subId = SubscriptionManager.getDefaultDataSubId();
-        getPhone(subId).setDataEnabled(false);
+        setDataEnabled(false);
         return true;
     }
 
@@ -1949,7 +1945,17 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
     @Override
     public void setDataEnabled(boolean enable) {
         enforceModifyPermission();
-        mPhone.setDataEnabled(enable);
+        long subId = SubscriptionManager.getDefaultDataSubId();
+        int phoneId = SubscriptionManager.getPhoneId(subId);
+
+        if (TelephonyManager.getDefault().isMultiSimEnabled()) {
+            // If the non-subid API is used, we want to update settings as well
+            // (see DataUsageSummary.java in packages/apps/Settings)
+            android.provider.Settings.Global.putInt(mPhone.getContext().getContentResolver(),
+                    android.provider.Settings.Global.MOBILE_DATA + phoneId, enable ? 1 : 0);
+        }
+
+        getPhone(phoneId).setDataEnabled(enable);
     }
 
     /**
@@ -1982,7 +1988,8 @@ public class PhoneInterfaceManager extends ITelephony.Stub {
             mApp.enforceCallingOrSelfPermission(android.Manifest.permission.MODIFY_PHONE_STATE,
                     null);
         }
-        return mPhone.getDataEnabled();
+        long subId = SubscriptionManager.getDefaultDataSubId();
+        return getPhone(subId).getDataEnabled();
     }
 
     @Override
-- 
1.9.3 (Apple Git-50)


From 550f3c9ec8a5fe0a98e0982a2c580f83615d999f Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Sun, 23 Nov 2014 15:57:20 +0700
Subject: [PATCH 4/5] Telephony: MSim: enable preferred mode button for all
 phones

HACK to allow changing network mode on multi-SIM

Change-Id: I79cd6dfb1653c9c1ef14b8965c9af2bd0f5792cb
---
 res/values/strings.xml                             | 10 ++++++++
 res/xml/msim_network_sub_setting.xml               |  4 ++--
 .../phone/msim/MSimMobileNetworkSubSettings.java   | 27 +++++++---------------
 3 files changed, 20 insertions(+), 21 deletions(-)

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 8fbc2eb..7719934 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -551,6 +551,16 @@
     <string name="preferred_network_mode_gsm_only_choice">GSM only</string>
     <string name="preferred_network_mode_gsm_wcdma_preferred_choice">GSM/WCDMA preferred</string>
 
+    <string-array name="preferred_network_mode_choices_2" translatable="false">
+        <item>@string/preferred_network_mode_wcdma_only_choice</item>
+        <item>@string/preferred_network_mode_gsm_only_choice</item>
+        <item>@string/preferred_network_mode_gsm_wcdma_preferred_choice</item>
+    </string-array>
+    <string-array name="preferred_network_mode_values_2"  translatable="false">
+        <item>"2"</item>
+        <item>"1"</item>
+        <item>"0"</item>
+    </string-array>
 
     <string-array name="preferred_network_mode_choices" translatable="false">
         <item>@string/preferred_network_mode_td_scdma_lte_gsm_wcdma_cdma_evdo_choice </item>
diff --git a/res/xml/msim_network_sub_setting.xml b/res/xml/msim_network_sub_setting.xml
index 6cd0eb5..72101de 100644
--- a/res/xml/msim_network_sub_setting.xml
+++ b/res/xml/msim_network_sub_setting.xml
@@ -40,8 +40,8 @@
         android:key="preferred_network_mode_key"
         android:title="@string/preferred_network_mode_title"
         android:summary="@string/preferred_network_mode_summary"
-        android:entries="@array/preferred_network_mode_choices"
-        android:entryValues="@array/preferred_network_mode_values"
+        android:entries="@array/preferred_network_mode_choices_2"
+        android:entryValues="@array/preferred_network_mode_values_2"
         android:dialogTitle="@string/preferred_network_mode_dialogtitle" />
 
 </PreferenceScreen>
diff --git a/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java b/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java
index 3a5a23c..a657981 100644
--- a/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java
+++ b/src/com/android/phone/msim/MSimMobileNetworkSubSettings.java
@@ -212,32 +212,21 @@ public class MSimMobileNetworkSubSettings extends PreferenceActivity
             mUPLMNPref.getIntent().putExtra(PhoneConstants.SUBSCRIPTION_KEY, mPhone.getPhoneId());
         }
 
-        boolean isLteOnCdma = mPhone.getLteOnCdmaMode() == PhoneConstants.LTE_ON_CDMA_TRUE;
-        if (getResources().getBoolean(R.bool.world_phone) == true) {
-            // set the listener for the mButtonPreferredNetworkMode list preference so we can issue
-            // change Preferred Network Mode.
-            mButtonPreferredNetworkMode.setOnPreferenceChangeListener(this);
+        // set the listener for the mButtonPreferredNetworkMode list preference so we can issue
+        // change Preferred Network Mode.
+        mButtonPreferredNetworkMode.setOnPreferenceChangeListener(this);
 
-            //Get the networkMode from Settings.System and displays it
-            int settingsNetworkMode = getPreferredNetworkMode();
-            mButtonPreferredNetworkMode.setValue(Integer.toString(settingsNetworkMode));
+        //Get the networkMode from Settings.System and displays it
+        int settingsNetworkMode = getPreferredNetworkMode();
+        mButtonPreferredNetworkMode.setValue(Integer.toString(settingsNetworkMode));
+
+        if (getResources().getBoolean(R.bool.world_phone) == true) {
             mCdmaOptions = new CdmaOptions(this, prefSet, mPhone);
             mGsmUmtsOptions = new GsmUmtsOptions(this, prefSet, mPhone.getPhoneId());
         } else {
-            if (!isLteOnCdma) {
-                prefSet.removePreference(mButtonPreferredNetworkMode);
-            }
             int phoneType = mPhone.getPhoneType();
             if (phoneType == PhoneConstants.PHONE_TYPE_CDMA) {
                 mCdmaOptions = new CdmaOptions(this, prefSet, mPhone);
-                if (isLteOnCdma) {
-                    mButtonPreferredNetworkMode.setOnPreferenceChangeListener(this);
-
-                    int settingsNetworkMode = getPreferredNetworkMode();
-                    mButtonPreferredNetworkMode.setValue(
-                            Integer.toString(settingsNetworkMode));
-                }
-
             } else if (phoneType == PhoneConstants.PHONE_TYPE_GSM) {
                 mGsmUmtsOptions = new GsmUmtsOptions(this, prefSet, mPhone.getPhoneId());
             } else {
-- 
1.9.3 (Apple Git-50)


From 32eca75fd8632797e2ff4d2c11773f66e7f79256 Mon Sep 17 00:00:00 2001
From: Pawit Pornkitprasan <p.pawit@gmail.com>
Date: Tue, 23 Dec 2014 22:27:09 +0700
Subject: [PATCH 5/5] Telephony: HACK: fix 'up' for MSimCallFeature sub option

Change-Id: I1fcd9309d2849fbdd0ac1c903cbda4a8b1f213f5
---
 src/com/android/phone/CallFeaturesSetting.java | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/com/android/phone/CallFeaturesSetting.java b/src/com/android/phone/CallFeaturesSetting.java
index 4cadc7a..492edba 100644
--- a/src/com/android/phone/CallFeaturesSetting.java
+++ b/src/com/android/phone/CallFeaturesSetting.java
@@ -2188,10 +2188,13 @@ public class CallFeaturesSetting extends PreferenceActivity
      * This is useful for implementing "HomeAsUp" capability for second-level Settings.
      */
     public static void goUpToTopLevelSetting(Activity activity) {
-        Intent intent = new Intent(activity, CallFeaturesSetting.class);
-        intent.setAction(Intent.ACTION_MAIN);
-        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
-        activity.startActivity(intent);
+        // HACK: See MSimMobileNetworkSubSettings#onOptionsItemSelected
+        if (!PhoneUtils.isMultiSimEnabled()) {
+            Intent intent = new Intent(activity, CallFeaturesSetting.class);
+            intent.setAction(Intent.ACTION_MAIN);
+            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
+            activity.startActivity(intent);
+        }
         activity.finish();
     }
 }
-- 
1.9.3 (Apple Git-50)

